### PR TITLE
App submission: monero

### PR DESCRIPTION
**Pull request by deverickapollo**
# App Submission

### 256x256 SVG icon
  - https://svgshare.com/i/vGT.svg

### Gallery images
  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Monero%20Dashboard%20Menu.png
  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Connect%20Wallet.png
  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Advanced%20Settings.png

### I have tested my app on:
- [x] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [x] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [x] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

File: monero/umbrel-app.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  This is a full node that will download the entire Monero blockchain and store it on your Umbrel. 
+  This is the most secure way to run a Monero node, but it will take a long time to sync and will require a lot of storage space.
+developer: Monero
+website: https://www.getmonero.org/
+dependencies: []
+repo: https://github.com/deverickapollo/umbrel-monero
+support: https://github.com/deverickapollo/umbrel-monero/issues
+port: 9976
+submitter: Deverick
+submission: https://github.com/hashes4merkle/hundredsats
```
Comment (m1m3-50) [inline monero/umbrel-app.yml:19]: redirects to https://github.com/deverickapollo/meganero - maybe it's better to have the actual link here?

File: monero/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        ipv4_address: $APP_MONERO_IP
+
+  monerod:
+    user: "1000:1000"
+    restart: unless-stopped
+    stop_grace_period: 15m
+    command: "${APP_MONERO_COMMAND}"
+    healthcheck:
+      disable: true
+    image: sethsimmons/simple-monerod:latest
```
Comment (m1m3-50) [inline monero/docker-compose.yml:51]: I believe it'd be better to use a most recent tag (v0.18.2.2 at the moment) instead of latest... @sethforprivacy, mind sharing your thoughts?

File: monero/umbrel-app.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+manifestVersion: 1.1
+id: meganero-monero
+category: Finance
+name: Monero Node
+icon: https://raw.githubusercontent.com/deverickapollo/umbrel-monero/dev/ui/src/assets/community-monero-icon.png
+version: "1.0.0"
```
Comment (m1m3-50) [inline monero/umbrel-app.yml:6]: why not use Monero node's version here?

**Review by m1m3-50 — CHANGES_REQUESTED** [review]
Please check my comments about versioning for the app itself and also the monero node image used.

File: monero/docker-compose.yml (hunk 2 above)
Comment (deverickapollo) [reply monero/docker-compose.yml:51]: Good call - I hadn't pinned versions yet.  I'll also push out a release for the frontend.

**Review by IMPranshu — CHANGES_REQUESTED** [review]

**Comment by deverickapollo** [comment]
Updates complete.  Thanks!

**Comment by nmfretz** [comment]
Apologies for the delay in reviewing this @deverickapollo! I have had my head down working on other parts of the code-base. I will try and review soon.

Thanks @m1m3-50 and @IMPranshu for stepping up and helping. Much appreciated.

File: monero/umbrel-app.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+manifestVersion: 1.1
+id: monero
+category: Finance
+name: Monero Node
+icon: https://raw.githubusercontent.com/deverickapollo/umbrel-monero/dev/ui/src/assets/community-monero-icon.png
+version: "0.18.2.2"
+tagline: Run your monero node and independently store and validate every single Monero transaction with it.
```
Comment (nmfretz) [inline monero/umbrel-app.yml:7]: Any ideas on a shorter tagline so that it can show up nicely in the app store? Maybe something as simple as `Run your personal Monero node`

image

File: monero/umbrel-app.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+submitter: Deverick
+submission: https://github.com/getumbrel/umbrel-apps/pull/690
+gallery:
+  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Monero%20Dashboard%20Menu.png
+  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Connect%20Wallet.png
+  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Advanced%20Settings.png
+path: ""
+defaultPassword: ""
+releaseNotes: >-
+ Initial Release
```
Comment (nmfretz) [inline monero/umbrel-app.yml:27]: For this initial release we can just leave this as 
```
releaseNotes: ""
```

and then they won't show in the UI, but the yaml will remain to remind us to update it in future app versions.

File: monero/umbrel-app.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+category: Finance
+name: Monero Node
+icon: https://raw.githubusercontent.com/deverickapollo/umbrel-monero/dev/ui/src/assets/community-monero-icon.png
+version: "0.18.2.2"
+tagline: Run your monero node and independently store and validate every single Monero transaction with it.
+description: >-
+  Run your monero node and independently store and validate every single Monero transaction with it.
+  This is a full node that will download the entire Monero blockchain and store it on your Umbrel. 
+  This is the most secure way to run a Monero node, but it will take a long time to sync and will require a lot of storage space.
+developer: Monero
```
Comment (nmfretz) [inline monero/umbrel-app.yml:12]: This should be you, woop!

File: monero/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      PORT: "9976"
+      MONERO_HOST: "${APP_MONERO_NODE_IP}"
+      MONERO_P2P_PORT: "${APP_MONERO_P2P_PORT}"
+      MONERO_RPC_PORT: "${APP_MONERO_RPC_PORT}"
+      MONERO_DEFAULT_NETWORK: "${MONERO_DEFAULT_NETWORK}"
+      MONERO_RPC_USER: "${APP_MONERO_RPC_USER}"
+      MONERO_RPC_PASSWORD: "${APP_MONERO_RPC_PASS}"
+      RPC_PORT: "${APP_MONERO_RPC_PORT}"
+      RPC_USER: "${APP_MONERO_RPC_USER}"
+      RPC_PASSWORD: "${APP_MONERO_RPC_PASS}"
```
Comment (nmfretz) [inline monero/docker-compose.yml:29]: I think these three env vars are actually legacy code that we need to remove from our Bitcoin Node app, so you may be able to remove these without issue.

You can see that they are essentially duplicates of:

```
MONERO_RPC_PORT: "${APP_MONERO_RPC_PORT}"
MONERO_RPC_USER: "${APP_MONERO_RPC_USER}"
MONERO_RPC_PASSWORD: "${APP_MONERO_RPC_PASS}"
```

File: monero/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    command: "${APP_MONERO_COMMAND}"
+    healthcheck:
+      disable: true
+    image: sethsimmons/simple-monerod:v0.18.2.2@sha256:835a9305b02c0a640363ddc2e77d709289ff54c402ede0a7ca242ab70b6270d4
+    ports:
+      - "${APP_MONERO_P2P_PORT}:${APP_MONERO_P2P_PORT}"
+      - "${APP_MONERO_RPC_PORT}:${APP_MONERO_RPC_PORT}"
+    expose:
+      - "${APP_MONERO_P2P_PORT}"
+      - "${APP_MONERO_RPC_PORT}"
```
Comment (nmfretz) [inline monero/docker-compose.yml:57]: Did you find that exposing the ports was necessary? I think you should be okay just binding the ports as you have done above.

File: monero/exports.sh [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	then
+	MONERO_INSTALL_EXISTS="true"
+	fi
+
+	APP_CONFIG_EXISTS="false"
+	if [[ -f "${EXPORTS_APP_DIR}/data/app/monero-config.json" ]]
+	then
+	APP_CONFIG_EXISTS="true"
+	fi
+} || true
```
Comment (nmfretz) [inline monero/exports.sh:111]: Are you using this block? This was created for the Bitcoin Node app when we introduced "Advanced Settings", where we didn't want to clobber existing user's manual settings. Since it looks like Advanced Settings has always been part of your Monero App, you can likely remove this and associated code in the forked umbrel-bitcoin repo.

File: monero/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+      default:
+        ipv4_address: $APP_MONERO_IP
+
+  monerod:
+    user: "1000:1000"
+    restart: unless-stopped
+    stop_grace_period: 15m
+    command: "${APP_MONERO_COMMAND}"
+    healthcheck:
+      disable: true
```
Comment (nmfretz) [inline monero/docker-compose.yml:50]: Is this to override a healthcheck designated in the Dockerfile that occurs super often?

File: monero/umbrel-app.yml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+dependencies: []
+repo: https://github.com/deverickapollo/umbrel-monero
+support: https://github.com/deverickapollo/umbrel-monero/issues
+port: 9976
+submitter: Deverick
+submission: https://github.com/getumbrel/umbrel-apps/pull/690
+gallery:
+  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Monero%20Dashboard%20Menu.png
+  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Connect%20Wallet.png
+  - https://raw.githubusercontent.com/deverickapollo/umbrel-monero/master/pictures/Advanced%20Settings.png
```
Comment (nmfretz) [inline monero/umbrel-app.yml:23]: Since this will be on the official app store, these gallery images will be hosted here: https://github.com/getumbrel/umbrel-apps-gallery

This section can be changed to simple numbered images. See bitcoin for example:

https://github.com/getumbrel/umbrel-apps/blob/5108c2641ed0164dc346b0e6032a09ee2aecdf09/bitcoin/umbrel-app.yml#L29-L32

**Review by nmfretz — COMMENTED** [review]
Excellent work on this @deverickapollo!  See below for some suggested changes. This is a fantastic addition for the Monero community.

**Comment by nmfretz** [comment]
I installed the app and am syncing the blockchain now. Initial sync is working great right off the bat. 

The primary purpose of my review is just to make sure that the app is packaged correctly for Umbrel; however, since I have some familiarity with the UI and server code of your app from working on our Bitcoin Node app, I can offer some additional help.

Check out this video below (no sound) where I highlight a few bugs I found while checking out the app. This is my first time playing around with monero, so apologies if I have misunderstood anything.

https://github.com/getumbrel/umbrel-apps/assets/85373263/fd5e03c2-2221-4621-b1bb-1389f59ca772

Notes the video:
- `Outgoing Connections to Tor Peers` is enabled in Advanced Settings, but there are no related lines in the bitmonero.conf. You can see in the UI that I have no connections to tor peers, and I confirmed this with RPC [`/get_connections`](https://www.getmonero.org/resources/developer-guides/daemon-rpc.html#get_connections). You likely need to update your `settingsToMultilineConfString()` function to write the required tor config to your bitmonero.conf
- The info popup next to the blockchain size is showing up even though `Prune Blocks` is disabled. This should be an easy conditional fix in the UI.
- Is I2P implemented in your app yet? monerod errors out when enabling `Outgoing Connections to I2P Peers`. If it's still a work in-progress, it could be worth removing the i2pd container and all associated code from this version so that users don't unwittingly crash their node and need to change both the `monero-config.json` and `bitmonero.conf` to get things back up and running.
- When I try to restart the node with `Incoming Connections` enabled in Advanced Settings, the UI shows an error and doesn't let me do it.
- If I disable `Outgoing Connections to Tor Peers` and try to restart the node, the UI shows an error. This error handling is included in our Bitcoin Node app to prevent users from disabling connections to **all** peers on clearnet, tor, and I2P. If they did this, then they would not be able to connect to any peers. However, it looks like there is no toggle to disable connections to clearnet peers in your app, so you just need to update the error logic in the UI to allow both Tor and I2P connections to be disabled at the same time.

This was just from a quick play, so there are probably other sneaky bugs that have understandably resulted from translating the bitcoin app to a monero app.

**Comment by nmfretz** [comment]
@deverickapollo I noticed that monerod uses quite a lot of RAM. This is during initial block download, with DB Sync Mode set to the `fast` default. I believe I have seen it go as high as 9GB for me (I'm testing with 16GB RAM). Do you know if monerod frees up RAM for other processes when needed?

```
CONTAINER ID   NAME                        CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O         PIDS
7fab4c46fd05   monero_monerod_1            171.84%   6.07GiB / 15.42GiB    39.37%    25.8GB / 427MB    28.1GB / 294GB    29
```

image

I have also noticed that if I try and restart my Umbrel while Monero is installed and running, it can take a very long time to shut down. Then when everything comes back online, the monero app UI can become stuck in this state even though monerod is plugging away in the background:

https://github.com/getumbrel/umbrel-apps/assets/85373263/546bb8a9-8f17-430c-8156-3ab6fc960e87

**Comment by deverickapollo** [comment]
@nmfretz - This is great! Thanks for taking the time to review the app!  A few of these issues I am aware of.  I should be able to address most in the next week.  I'll reach out once that work is complete.

**Comment by nmfretz** [comment]
Excellent, looking forward to it @deverickapollo! I have a monero node happily plugging away on one of my Umbrels, so I can help with any testing you need.

**Comment by ceramicwhite** [comment]
I'm running this. Fully synced it's sitting at <1gb ram and during syncing I never saw it go above 4gb

File: monero/exports.sh (hunk 9 above)
Comment (deverickapollo) [reply monero/exports.sh]: Removing and resolving as dead code. Thanks!

**Comment by nmfretz** [comment]
@deverickapollo let me know whenever you are ready for me to continue reviewing so I don't accidentally hold you up!

**Comment by ParthJadhav** [comment]
Hey @deverickapollo , just following up on the PR. Is it ready to be reviewed?

File: monero/docker-compose.yml (hunk 10 above)
Comment (deverickapollo) [reply monero/docker-compose.yml:47]: Looking into this now - I believe this is related to some work I'm doing to prepare monerod for an integration with btcpay.  I believe we can remove this for now.

**Comment by oscoDOTblog** [comment]
> Just wanted to give you a heads up. I really hope that Monero shitcoinery won't make its way onto the UmbrelOS App Store. If it does, i will quit, as iam sure many will. **Bitcoin-only.**

🤣 🤣 🤣 🤣 🤣 

You will be missed

**Comment by m1m3-50** [comment]
> Just wanted to give you a heads up. I really hope that Monero shitcoinery won't make its way onto the UmbrelOS App Store. If it does, i will quit, as iam sure many will. **Bitcoin-only.**

closed minded doesn't mix well with open source, you know...

**Comment by deverickapollo** [comment]
@nmfretz @ParthJadhav - Sorry for the hold up here. I addressed most of the highlighed concerns in the latest version published. Overall, I think we're good to ho here but give me a shout if the RAM issues or UI issues are blocking. I can confirm monerod RPC does seem a bit sluggish while completing IBD.  I'll continue to look into that.

**Comment by nmfretz** [comment]
Excellent, thanks @deverickapollo. I will check this out again this week!

**Comment by deverickapollo** [comment]
@nmfretz - how's this looking?

**Comment by nmfretz** [comment]
Thanks for the reminder @deverickapollo! Looking now.

**Comment by nmfretz** [comment]
@deverickapollo I am currently testing this out and monerod is showing that I am actively syncing. I am at around 8%, but the frontend is still not showing activity. The main `status` fetch is still returning `{operational: false}`. Have you noticed issues with the frontend RPC calls to monero during sync?

File: monero/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    networks:
+      default:
+        ipv4_address: $APP_MONERO_IP
+
+  monerod:
+    user: "1000:1000"
+    restart: unless-stopped
+    stop_grace_period: 15m
+    command: "${APP_MONERO_COMMAND}"
+    image: sethsimmons/simple-monerod:v0.18.3.1@sha256:1ef52fe371a8085c1cf8ba579610f0a33537a170561f2f1370d4f528055f170f
```
Comment (nmfretz) [inline monero/docker-compose.yml:46]: This just needs to be changed to the multi-arch digest so it works on arm64 and amd64. I get this:

```
$ sudo docker pull sethsimmons/simple-monerod:v0.18.3.1
v0.18.3.1: Pulling from sethsimmons/simple-monerod
Digest: sha256:294f8460d63240ba199e63c52cbd01ab71fbd286c2af62a19ae8a73a302afabd
```

**Review by nmfretz — COMMENTED** [review]
A couple changes to make

**Comment by nmfretz** [comment]
Also, can you please remove the Umbrel logo from the loading screen since it is an app made by Umbrel? I believe the relevant code is here:
https://github.com/deverickapollo/umbrel-monero/blob/master/ui/src/components/Loading.vue

image

**Comment by nmfretz** [comment]
@deverickapollo do you happen to know if RPC unresponsiveness during early sync is due to an underlying issue in monerod, the config, the app server, or a combination of all?

**Comment by nmfretz** [comment]
Also, @deverickapollo we don't want to hold you up for things that aren't actual blockers for this going live on the app store. There is already a large number of people running this successfully on Umbrel via a Community App Store, and you'll obviously make updates to this app over time to tweak performance.

What do you think about launching this now as-is, with some wording in the app description along the line of:

"Please note that after installing Monero Node, the app interface may not display any sync progress for several hours. This behaviour is typical at this stage of development. The initial sync is still happening in the background, and you'll see the status reflected in the app as soon as it's available.""

^ Please feel free not to use this, I'm just spitballing here.

**Comment by nmfretz** [comment]
@deverickapollo any progress or thoughts on this?

**Comment by grigio** [comment]
any updates?

**Comment by deverickapollo** [comment]
@nmfretz -  Bumped the app to 1.0.4 which should handle the startup issues noted. Some of the peering nodes defined in the config file were offline and caused the latency during boot. Let me know if you're still experiencing any issues there. If twe're good there, I am ok with shipping the current version.

**Comment by nmfretz** [comment]
Thanks @deverickapollo, I just tested now and I still have the same issue noted [here](https://github.com/getumbrel/umbrel-apps/pull/690#issuecomment-1866965825) where monerod isn't accepting rpc calls from the UI during the first X% of IBD.

When I view the monerod logs I see that sync starts right away and progresses normally. It's just that the frontend UI can't get any info back from monerod for quite a while into sync. I tested rpc calls from inside the monerod container and they work fine during initial sync.

What is your preference for launching now with this in the app description:
```
Please note that after installing Monero Node, the app interface may not display any sync progress for several hours.
This behaviour is typical at this stage of development. The initial sync is still happening in the background,
and you'll see the status reflected in the app as soon as it's available.
```

versus continuing to troubleshoot?

**Comment by deverickapollo** [comment]
@nmfretz - That should be fine. I'll investigate those issue and update in a later release.

**Comment by tylerwillis** [comment]
Any update on when to expect this in the official App Store? 

I got this running via community app store, but would love to recommend it to friends without signing up to be their IT Support on how to use community app stores. :)

**Comment by nmfretz** [comment]
Hey @deverickapollo I thought I had posted this here, but it appears I never did. Sorry about that.

When I was testing last month, I fully synced Monero and noticed that the UI shows 0% synced when monerod is on the latest block. This results in the loading animation permanently running in the Blockchain card and could be confusing to most users. Is this the case for you?

Screenshot 2024-03-03 at 8 40 59 PM

**Comment by deverickapollo** [comment]
@nmfretz Funny you mention it - I was just playing around with this last night. It turns out once the node is synced, target_height goes to 0 [Issue 7029](https://github.com/monero-project/monero/issues/7029]). I was using target_height to track our sync status and some of my recent changes didn't consider this. I should have a fix ready to go soon.

**Comment by sethforprivacy** [comment]
Please note that I'll be shutting down my Docker Hub account at some point in the near future, so GHCR should be used for my image here instead, i.e. for v0.18.3.3:

`ghcr.io/sethforprivacy/simple-monerod:v0.18.3.3`

**Comment by deverickapollo** [comment]
Appreciate the heads up! I'll get that change in asap

**Comment by nmfretz** [comment]
Thanks a bunch for noting that @sethforprivacy 🤝

**Comment by deverickapollo** [comment]
Monerod version change complete. Fix for sync % displayed on startup and full sync also added

**Comment by markisoke** [comment]
@deverickapollo Any way you will have the option to enable ZMQ in your app for us P2Pool users?

**Comment by nmfretz** [comment]
Thanks @deverickapollo. I still get the same original issue where sync percent doesn't show in the UI:

image

Can you confirm whether you have this same issue?

**Comment by highghlow** [comment]
+1 for this. Monero is the best crypto and a dedicated server is the best way to sync it

**Comment by highghlow** [comment]
> Thanks @ deverickapollo. I still get the same original issue where sync percent doesn't show in the UI:
> Can you confirm whether you have this same issue?

I've tested the app myself and can confirm that the issue is present. While the node is syncing, no progress gets displayed

**Comment by deverickapollo** [comment]
@highghlow - thanks for confirming. 

I'll review this asap. I just completed a refactor on the backend to replace monero-javascript with monero-ts so I'll include those changes and retest.

**Comment by nmfretz** [comment]
Thanks @deverickapollo, ping me when you're ready!

**Comment by highghlow** [comment]
Update: it updates the sync progress when Umbrel is rebooted

**Comment by nmfretz** [comment]
Hey @deverickapollo, out of curiosity I did a deep dive into this issue this morning. Hopefully this will help you with a fix!

**TLDR:**

1) Keep at least `--rpc-bind-ip=0.0.0.0` in the monerod start command so that the external server container can make RPC calls even before your custom bitmonero.conf is sourced.
2) Make sure that the server can restart monerod on initial boot so that the bitmonero.conf is sourced on app install. Right now, it fails to do so even with Fix (1) noted above.

After that this is good to go!

### 1) First, the general RPC issue where the UI cannot make any calls to monerod:

In version 1.0.1 you removed [these cli arguments](https://github.com/deverickapollo/umbrel-apps/blob/97cd78f44598d3c803e1f0e2c5050e3c8ff7cc1f/monero/exports.sh#L81-L85) that used to be passed to the monerod command on container start, and placed their equivalents in the logic that creates the default bitmonero.conf:

From exports.sh: https://github.com/deverickapollo/umbrel-apps/commit/aa48aaab50ee2c7b4142e31a210e73e13a5f91ee
```
BIN_ARGS=()
# Commenting out options that are replaced by generated config file. We should migrate all these over in a future update.
# BIN_ARGS+=( "--rpc-bind-port=${APP_MONERO_RPC_PORT}" )
# BIN_ARGS+=( "--rpc-bind-ip=0.0.0.0" )
# BIN_ARGS+=( "--confirm-external-bind" )
```

From umbrel-monero logic/disk.js: https://github.com/deverickapollo/umbrel-monero/commit/2c612fb511047b99d7630ff1320b4385a2c66436
```
  umbrelMoneroConfig.push(`rpc-bind-ip=0.0.0.0`);
  umbrelMoneroConfig.push(`rpc-restricted-bind-ip=0.0.0.0`);
  umbrelMoneroConfig.push(`rpc-bind-port=${constants.MONERO_RPC_PORT}`);
  umbrelMoneroConfig.push(`rpc-restricted-bind-port=${constants.MONERO_RESTRICTED_RPC_PORT}`);
  ```
  
This change causes issues because there isn't actually any bitmonero.conf when the app is first installed. It's created by the server container after initial boot, so without a custom bitmonero.conf or `rpc-bind-ip` cli argument, monerod actually defaults to binding 127.0.0.1 instead of 0.0.0.0. This means that applications outside of the monerod container (like the server container) can't make RPC calls. You won't see this issue when testing locally on your computer because the [docker-compose.yml file in the umbrel-monero repo](https://github.com/deverickapollo/umbrel-monero/blob/b4cf2c3a1a335f59a625da99ef12fb78f4d67c2d/docker-compose.yml#L37) binds the RPC port to 0.0.0.0 in the monerod command.

There is logic in the server container to [restart monerod when the bitmonero.conf file is created](https://github.com/deverickapollo/umbrel-monero/blob/b4cf2c3a1a335f59a625da99ef12fb78f4d67c2d/bin/www#L40-L59) (see next point below for a second issue), but without rpc-bind-ip=0.0.0.0 on initial boot of monerod, the server can't actually send an RPC call to tell monerod to restart.

@highghlow is right that when you restart umbrel, the UI will then start showing sync progress. And actually it's just that when you restart the monerod container, progress will be shown, because monerod will then source the bitmonero.conf that the server container created. You can test this by installing monero and seeing that no RPC calls can be made externally. Then run: 
`sudo docker restart monero_monerod_1`
to restart the container. monerod will then source the bitmonero.conf and the UI will show progress.

### 2) Even when starting monerod with rpc-bind-ip=0.0.0.0, the server container doesn't seem to be able to restart monerod when it is first booting up.

Once things have settled, you can restart monerod from Advanced Settings, but when the app is first installed you will see this error:

```
> umbrel-middleware@0.4.0 start
> node ./bin/www

Listening on port 9976
bitmonero.conf does not exist, creating config files with default values
MoneroError: RequestError: Error: connect ECONNREFUSED 10.21.21.179:18081
    at MoneroRpcConnection.sendPathRequest (/app/node_modules/monero-javascript/src/main/js/common/MoneroRpcConnection.js:320:18)
    at async MoneroDaemonRpc.stop (/app/node_modules/monero-javascript/src/main/js/daemon/MoneroDaemonRpc.js:785:16)
    at async onmessage (/app/node_modules/monero-javascript/src/main/js/common/MoneroWebWorker.js:44:49) {
  code: undefined
}
```

The server will keep trying to [restart monerod for 1 minute](https://github.com/deverickapollo/umbrel-monero/blob/b4cf2c3a1a335f59a625da99ef12fb78f4d67c2d/bin/www#L40-L59) in order to source the newly created (or changed by a future app update) bitmonero.conf:

bin/www:
```
async function createConfFilesAndRestartMonerod() {
  console.log('bitmonero.conf does not exist, creating config files with default values');
  const config = await diskLogic.getJsonStore();

  await diskLogic.applyCustomMoneroConfig(config, true);

  const MAX_TRIES = 60;
  let tries = 0;

  while (tries < MAX_TRIES) {
    try {
      await monerodLogic.stop();
      break;
    } catch (error) {
      console.error(error);
      tries++;
      await new Promise((resolve) => setTimeout(resolve, 1000));
    }
  }
}
```

I haven't explored this one further, but it is possible that monerod just isn't ready to accept RPC requests this early on in the initial boot. Without this initial restart, users will be running monerod with default config values and not inheriting anything you have set as your initial bitmonero.conf. They would only source the bitmonero.conf if they restart the app, restart umbrel, or change settings and restart through Advanced Settings within the monero app.

**Comment by nmfretz** [comment]
Hey @deverickapollo ping me when you are ready for me to test!

**Comment by deverickapollo** [comment]
@nmfretz Appreciate that additional input! I was able to reproduce this issue and identify the problem. I've included a fix in the latest release that should address the startup issue.

**Comment by nmfretz** [comment]
Well done @deverickapollo! This is good to go from a packaging perspective. Welcome to the official app store 🎉 

image

**Comment by oromanobr** [comment]
hi. no more updatem in umbrel?

**Comment by al-lac** [comment]
Hey @oromanobr, the last update was pushed about a month ago, are you missing some update?
